src/DotNetty.Handlers/Timeout/WriteTimeoutHandler.cs

```diff
@@ -0,0 +1,199 @@
+        /// Initializes a new instance of the <see cref="DotNetty.Handlers.Timeout.ReadTimeoutHandler"/> class.
+        /// </summary>
+        /// <param name="timeout">Timeout.</param>
+        public WriteTimeoutHandler(TimeSpan timeout)
+        {
+            this.timeout = 
+                timeout.Ticks <= 0 ? MinTimeout : TimeUtil.Max(timeout, MinTimeout);
+        }
+
+        public override Task WriteAsync(IChannelHandlerContext context, object message)
+        {
+            Task task = context.WriteAsync(message);
+
+            if (this.timeout.Ticks > 0)
+            {
+                this.ScheduleTimeout(context, task);
+            }
+
+            return task;
+        }
+
+        public override void HandlerRemoved(IChannelHandlerContext context)
+        {
+            LinkedListNode<WriteTimeoutTask> task = this.tasks.Last;
+            while (task != null)
+            {
+                task.Value.ScheduledTask.Cancel();
+                this.tasks.RemoveLast();
+                task = this.tasks.Last;
+            }
+        }
+
+        void ScheduleTimeout(IChannelHandlerContext context, Task promise)
+        {
+            var task = new WriteTimeoutTask(context, promise, this);
+            var wrappedTask = new LinkedListNode<WriteTimeoutTask>(task);
+
+            task.ScheduledTask = context.Executor.Schedule(task, timeout);
+
+            if (!task.ScheduledTask.Completion.IsCompleted)
+            {
+                this.AddWriteTimeoutTask(wrappedTask);
+            }
+        }
+
+        void AddWriteTimeoutTask(LinkedListNode<WriteTimeoutTask> task)
+        {
+            if (this.tasks.First == null)
+            {
+                this.tasks.AddFirst(task);
+            }
+            else
+            {
+                this.tasks.AddLast(task);    
+            }
+        }
+
+        void RemoveWriteTimeoutTask(WriteTimeoutTask task)
+        {
+            this.tasks.Remove(task);
+        }
+
+        /// <summary>
+        /// Is called when a write timeout was detected
+        /// </summary>
+        /// <param name="context">Context.</param>
+        protected void WriteTimedOut(IChannelHandlerContext context)
+        {
+            if (!this.closed)
+            {
+                context.FireExceptionCaught(WriteTimeoutException.Instance);
+                context.CloseAsync();
+                this.closed = true;
+            }
+        }
+
+        sealed class WriteTimeoutTask : OneTimeTask
+        {
+            readonly WriteTimeoutHandler handler;
+            readonly IChannelHandlerContext context;
+            readonly Task promise;
+
+            public WriteTimeoutTask(IChannelHandlerContext context, Task promise, WriteTimeoutHandler handler)
+            {
+                this.context = context;
+                this.promise = promise;
+                this.handler = handler;
+
+                promise.ContinueWith(new Action<Task>(this.HandleOperationComplete)); 
```

Comment:
use `ContinueWith(Action<Task, object> action, object state)` overload, pass `this` as a state, add a `static readonly Action<Task, object> OperationCompleteAction = HandleOperationComplete;` and derive `WriteTimeoutTask` out of `state` in `HandleOperationComplete`.